### PR TITLE
4672: Return all solutions when no capabilities selected

### PR DESCRIPTION
**Pull request by kat1906**

### CODE REVIEW

File: app/routes.js [hunk 1]
```diff
@@ -42,9 +42,11 @@ router.get('/solutions', (req, res) => {
 
 router.post('/solutions/capabilities-selector', withCatch(async (req, res) => {
   let capabilitiesSelected = req.body.capabilities;
+  if (!capabilitiesSelected) capabilitiesSelected = 'all';
   if (!Array.isArray(capabilitiesSelected)) capabilitiesSelected = [capabilitiesSelected];
-  const capabilitiesParam = capabilitiesSelected.join('+');
-  res.redirect(`/solutions/capabilities-selector.${capabilitiesParam}`);
+  const capabilitiesParam = `.${capabilitiesSelected.join('+')}`;
```
Comment (alibombaye) [inline app/routes.js]: can we move the above lines to a helper function that works out the capabilitiesParam. Might also get rid of that `let`

File: app/routes.test.js [hunk 2]
```diff
@@ -216,17 +216,48 @@ describe('routes', () => {
...
+        .mockImplementation(() => Promise.resolve(mockFilteredSolutions));
+      const app = new App().createApp();
+      app.use('/', routes);
+      return request(app)
+        .post('/solutions/capabilities-selector')
+        .send({ capabilities: 'C1' })
+        .expect(302)
+        .then((res) => {
+          expect(res.redirect).toEqual(true);
+          expect(res.text).toEqual('Found. Redirecting to /solutions/capabilities-selector.C1');
```
Comment (alibombaye) [inline app/routes.test.js]: I don't get this? is the response we return on the page?

Comment (kat1906) [reply app/routes.test.js]: it's just a key added to the res object - I was looking for where I could find the redirect url to test and found this which has it in

Comment (alibombaye) [reply app/routes.test.js]: ok so this is not the actual body of the res?

Comment (kat1906) [reply app/routes.test.js]: I have no idea what you mean but it is put on the res by the redirect so there is not much we can do about it if you don't like it 😄

Comment (alibombaye) [reply app/routes.test.js]: haha why so defensive. I like it! I understand why now. The result of the unit test is the redirect rather an actual page to render. That's what confused me

File: app/routes.js (hunk 1 above)
Comment (kat1906) [reply app/routes.js]: I've moved it but there is still a let because the way the value is worked out is complex

File: app/routes.test.js (hunk 2 above)
Comment (kat1906) [reply app/routes.test.js]: yeah I just wanted to access the redirect url and this was the first place I found it

File: app/helpers/routerHelper.test.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
...
+
+    it('returns the correct param string if one capability is sent', () => {
+      expect(getCapabilitiesParam('C1')).toEqual('.C1');
+      expect(getCapabilitiesParam('C14')).toEqual('.C14');
+    });
+
+    it('returns the correct param string if many capabilities are sent', () => {
+      expect(getCapabilitiesParam(['C1', 'C2', 'C3'])).toEqual('.C1+C2+C3');
+      expect(getCapabilitiesParam(['C14', 'C21', 'C7'])).toEqual('.C14+C21+C7');
+    });
```
Comment (alibombaye) [inline app/helpers/routerHelper.test.js]: 👍

Comment (kat1906) [reply app/helpers/routerHelper.test.js]: thought you would appreciate this file 😄

File: app/routes.test.js [hunk 4]
```diff
@@ -216,17 +216,48 @@ describe('routes', () => {
...
+    });
+
+    it('should return the correct status and text if there is no error and many capabilities selected', () => {
+      solutionListPageContext.getSolutionsForSelectedCapabilities = jest.fn()
+        .mockImplementation(() => Promise.resolve(mockFilteredSolutions));
+      const app = new App().createApp();
+      app.use('/', routes);
+      return request(app)
+        .post('/solutions/capabilities-selector')
+        .send({ capabilities: 'C1 C2 C3' })
```
Comment (alibombaye) [inline app/routes.test.js]: Should this be passed in as ['C1', 'C2', 'C3']?

File: app/routes.test.js [hunk 5]
```diff
@@ -216,17 +216,48 @@ describe('routes', () => {
...
+        .expect(302)
+        .then((res) => {
+          expect(res.redirect).toEqual(true);
+          expect(res.text).toEqual('Found. Redirecting to /solutions/capabilities-selector.C1%20C2%20C3');
```
Comment (alibombaye) [inline app/routes.test.js]: should these be separated by `+`?

File: app/routes.test.js (hunk 4 above)
Comment (kat1906) [reply app/routes.test.js]: not to our post endpoint - it comes in this format from the form submit

File: app/helpers/routerHelper.test.js (hunk 3 above)
Comment (alibombaye) [reply app/helpers/routerHelper.test.js]: yeah I have done. As a result it's made me understand and make more comments 😂

File: app/routes.test.js (hunk 4 above)
Comment (alibombaye) [reply app/routes.test.js]: don't get it. How else does this come to our post endpoint if not the form submit?

Comment (kat1906) [reply app/routes.test.js]: it comes in this format from the form submit

Comment (kat1906) [reply app/routes.test.js]: which bit is confusing you?

File: app/routes.test.js (hunk 5 above)
Comment (kat1906) [reply app/routes.test.js]: this is uri encoded - there was a bit in that article Sam posted about url encoding so this might change

**Pull request merged**